packages/hardhat-core/src/internal/artifacts.ts

```diff
@@ -372,6 +385,192 @@ export class Artifacts implements IArtifacts {
+  }
+
+  private _handleWrongArtifactForFullyQualifiedName(
+    fullyQualifiedName: string
+  ): never {
+    const names = this._getAllFullyQualifiedNamesSync();
+
+    if (names.length === 0) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        fullyQualifiedName,
+      });
+    }
+
+    const similarNames = this._getSimilarContractNames(
+      fullyQualifiedName,
+      names
+    );
+
+    throw new HardhatError(ERRORS.ARTIFACTS.TYPO_SUGGESTION, {
+      contractName: fullyQualifiedName,
+      similarNames: similarNames.join(os.EOL),
+    });
+  }
+
+  private _handleWrongArtifactForContractName(
+    contractName: string,
+    files: string[]
+  ): never {
+    const names = this._getAllContractNamesFromFiles(files);
+
+    if (names.length === 0) {
+      throw new HardhatError(ERRORS.ARTIFACTS.NOT_FOUND, {
+        contractName,
+      });
+    }
+
+    const similarNames = this._getSimilarContractNames(contractName, names);
+
+    if (similarNames.length === 1) {
+      throw new HardhatError(ERRORS.ARTIFACTS.TYPO_SUGGESTION, {
+        contractName,
+        similarNames: similarNames[0],
+      });
+    }
+
+    this._handleMultipleSimilarContractNames(contractName, files, similarNames);
+  }
+
+  /**
+   * cases:
+   *  given: 'Greter'
+   *    'contracts/Greeter.sol:Greeter'
+   *    'contracts/Meeter.sol:Greeter'
+   *    'Greater'
+   *
+   *  'Greeter', 'Greeter', and 'Greater' are all 1 edit-distance away from 'Greter'
+   *  because user gave contract name here, we want to display fqn's for duplicates in `similarNames`
+   */
+  private _handleMultipleSimilarContractNames(
+    contractName: string,
+    files: string[],
+    similarNames: string[]
+  ): never {
+    const outputNames = [];
+    const groups = similarNames.reduce((obj, cur) => {
+      obj[cur] = obj[cur] ? obj[cur] + 1 : 1;
+      return obj;
+    }, {} as { [k: string]: number });
+
+    for (const [name, occurrances] of Object.entries(groups)) {
+      if (occurrances > 1) {
+        for (const file of files) {
+          if (path.basename(file) === `${name}.json`) {
+            outputNames.push(
+              path.normalize(this._getFullyQualifiedNameFromPath(file))
+            );
+          }
+        }
+        continue;
+      }
+
+      outputNames.push(name);
+    }
+
+    throw new HardhatError(ERRORS.ARTIFACTS.TYPO_SUGGESTION, {
+      contractName,
+      similarNames: outputNames.join(os.EOL),
+    });
+  }
+
+  /**
+   *
+   * @param givenName can be FQN or contract name
+   * @param names MUST match type of givenName (i.e. array of FQN's if givenName is FQN)
+   * @returns
+   */
+  private _getSimilarContractNames(
+    givenName: string,
+    names: string[]
+  ): string[] {
+    let shortestDistance;
+    let mostSimilarNames: string[] = [];
+    for (const name of names) {
+      const distance = findDistance(givenName, name);
+
+      if (shortestDistance === undefined) {
```

Comment:
i agree with @fvictorio's suggestion. We should get a threshold and show everything under it.

Reply:
> The problem with this approach is that it will always suggest something:

I designed it that way intentionally, so I guess it's sort of a bug vs. feature discussion. Obviously the example you gave is exaggerated and wouldn't happen in a real use case; that being said, I did entertain the idea of a distance threshold like you suggested. 

I ran into a similar issue though where any number I picked seemed arbitrary, so I decided the most logical option would be just to not implement the threshold at this time and if it's something we see negative feedback about then it's simple to add in later.

Reply:
My example was an exaggeration but the idea that someone could use an artifact name that isn't close to any of the existing ones is not that far-fetched. Maybe someone renames a contract and forgets about it, or mis-remembers some name. In any case, the problem as I see it is that suggesting a name that it's not even close to the one that was entered is an unpolished experience.

> I ran into a similar issue though where any number I picked seemed arbitrary, so I decided the most logical option would be just to not implement the threshold at this time

I don't mind using an arbitrary number. I would use 3, just because it feels sensible, but I'd be ok with 2 or 4 too. This is a best-effort thing, so anything will be better than nothing.

Reply:
Alright I'll make that change